tests/test_engine.py

```diff
@@ -1131,6 +1131,52 @@ def flaky_function():
                     log_prints=False,
                 )
 
+    async def test_is_retriable_after_failure(
```

Comment:
Can we also add a test to ensure that when retries are set to 0, we do not retry regardless of whether it's retriable by default.

And after adding the check for `is_failed` for setting the `terminal_state.state_details.retriable` bool, I think it would make sense to add a test to ensure when a task completes, that we assert the `check_task_failure_retriable` is not called.

Reply:
1. Test added to test_engine.py to make sure the task does not retry when retries=0 even if retriable.
2. `check_task_failure_retriable` is only called when both are true:
    1. `terminal_state` is a failed state
    2. `task.retry_condition_fn` is a callable

Reply:
The tests look really good!